### DISCUSSION TITLE
SkywalkingAgent-5-GRPCChannelManager-0 GRPCChannelManager : Create channel to ${SKYWALKING_SERVER} fail.

### DISCUSSION THREAD

**Discussion by tubushitu** [apache/skywalking#10330] [General]
ERROR 2023-01-18 20:07:55:563 SkywalkingAgent-5-GRPCChannelManager-0 GRPCChannelManager : Create channel to ${SKYWALKING_SERVER} fail.
java.lang.ArrayIndexOutOfBoundsException: Index 1 out of bounds for length 1
at org.apache.skywalking.apm.agent.core.remote.GRPCChannelManager.run(GRPCChannelManager.java:133)
at org.apache.skywalking.apm.util.RunnableWithExceptionProtection.run(RunnableWithExceptionProtection.java:33)
at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Unknown Source)
at java.base/java.util.concurrent.FutureTask.runAndReset(Unknown Source)
at java.base/java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(Unknown Source)
at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(Unknown Source)
at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(Unknown Source)
at java.base/java.lang.Thread.run(Unknown Source)

I have encountered the same problem as yours, but my skywalking-Dskywalking.collector.backend_ Service=${SKYWALKING_SERVER} I passed it in the form of variables. The agent is in the container. I made the basic image of the docker

**Comment by wu-sheng** [comment]
Please follow document, and system property doesn't support system env variable.

**Reply by wu-sheng** [reply]
If you want to set through system env variable, use that directly.
Mixing system property and env variable is not working.

**Reply by tubushitu** [reply]
ERROR 2023-02-02 14:22:23:284 SkywalkingAgent-5-GRPCChannelManager-0 GRPCChannelManager : Create channel to ${sky} fail.
java.lang.ArrayIndexOutOfBoundsException: Index 1 out of bounds for length 1
        at org.apache.skywalking.apm.agent.core.remote.GRPCChannelManager.run(GRPCChannelManager.java:133)
        at org.apache.skywalking.apm.util.RunnableWithExceptionProtection.run(RunnableWithExceptionProtection.java:33)
        at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Unknown Source)
        at java.base/java.util.concurrent.FutureTask.runAndReset(Unknown Source)
        at java.base/java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(Unknown Source)
        at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(Unknown Source)
        at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(Unknown Source)
        at java.base/java.lang.Thread.run(Unknown Source)

FROM cnbjdocker.de.bj.cn.ao.ericsson.se/proj-rvc/skywalking:8.9.1
ARG sky=${sky}
ENV sky=${sky}
WORKDIR /home/
COPY   *.jar app.jar
EXPOSE 8080
CMD  [ "-javaagent:/usr/local/skywalking/agent/skywalking-agent.jar","-Dskywalking.agent.service_name=project-capacity-management-RVC-6","-Dskywalking.collector.backend_service=${sky}","-Dspring.cloud.nacos.discovery.server-addr=xx.xx.xx.xx30004","-Dspring.cloud.nacos.config.server-addr=xx.xx.xx.xx30004","-Dspring.cloud.nacos.config.enabled=true","-Dspring.cloud.nacos.config.refresh-enabled=true","-Dspring.cloud.nacos.discovery.enabled=true","-Dspring.cloud.nacos.discovery.instance-enabled=true","-Dspring.cloud.nacos.discovery.cluster-name=gateway-RVC962","-Dspring.application.name=gateway","-Dspring.cloud.inetutils.preferredNetworks=10.0.2","-jar","app.jar",">gateway.log" ]
ENTRYPOINT [ "java" ]

root@9b5e9e19d19b:/usr/local/skywalking/agent/logs# env
HOSTNAME=9b5e9e19d19b
JAVA_HOME=/usr/local/openjdk-11
PWD=/usr/local/skywalking/agent/logs
HOME=/root
LANG=C.UTF-8
SW_AGENT_NAME=RVC-7-gateway
sky=oap:11800
TERM=xterm
SHLVL=1
PATH=/usr/local/openjdk-11/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin
JAVA_VERSION=11.0.14.1
SW_AGENT_COLLECTOR_BACKEND_SERVICES=
OLDPWD=/usr/local/skywalking/agent
_=/usr/bin/env



I passed in the variable sky, but it doesn't seem to be recognized

Can you help me

**Reply by wu-sheng** [reply]
If you never read what I said. I can't see any point to follow this.

**Reply by tubushitu** [reply]
> If you want to set through system env variable, use that directly. Mixing system property and env variable is not working.

I don't understand the meaning of your sentence. I should use environment variables. I can also drop the environment variables I use in the container, but I don't seem to recognize them here. In addition, you said that the system environment variables are not supported. I tried to replace them but didn't recognize them

**Reply by wu-sheng** [reply]
Read docs, don't guess.
We never said system env is defined and used like this.

**Comment by Superskyyy** [comment]
If I understand correctly you are trying to pass your environment variable through system properties, but they are not automatically expanded by magic. So what you passed is a literal string `${sky}` not `oap:11800`. It doesn't really make sense when you can directly pass env variables without -D. 

@tubushitu

**Reply by tubushitu** [reply]
> If I understand correctly you are trying to pass your environment variable through system properties, but they are not automatically expanded by magic. So what you passed is a literal string `${sky}` not `oap:11800`. It doesn't really make sense when you can directly pass env variables without -D.
> 
> @Superskyyy
I am very grateful for your help. I am carefully analyzing what you mean,
I read the official document again, but I can't understand it


If the environment variable exists in your operating system and its value is , then the value of here will be overwritten to , otherwise, it will be set to .SW_AGENT_NAME skywalking-agent-demoagent.service_nameskywalking-agent-demoYour_ApplicationName

By the way, Placeholder nesting is also supported, like . In this case, if the environment variable not exists, but the environment variable exists and its value is , then the value of here will be overwritten to ,otherwise, it will be set to .${SW_AGENT_NAME:${ANOTHER_AGENT_NAME:Your_ApplicationName}}SW_AGENT_NAME ANOTHER_AGENT_NAMEskywalking-agent-demoagent.service_nameskywalking-agent-demoYour_ApplicationName


I still don't understand

**Reply by tubushitu** [reply]
> If I understand correctly you are trying to pass your environment variable through system properties, but they are not automatically expanded by magic. So what you passed is a literal string `${sky}` not `oap:11800`. It doesn't really make sense when you can directly pass env variables without -D.
> 
> @tubushitu

You mean I change the  collector.backend_service=${SW_AGENT_COLLECTOR_BACKEND_SERVICES:127.0.0.1:11800}    change   collector.backend_service=${SW_AGENT_COLLECTOR_BACKEND_SERVICES:${sky}}

dockerfile ：CMD [ "-javaagent:/usr/local/skywalking/agent/skywalking-agent.jar","-Dspring.cloud.nacos.discovery.server-addr=xx.xx.xx.xx30004","-Dspring.cloud.nacos.config.server-addr=xx.xx.xx.xx30004","-Dspring.cloud.nacos.config.enabled=true","-Dspring.cloud.nacos.config.refresh-enabled=true","-Dspring.cloud.nacos.discovery.enabled=true","-Dspring.cloud.nacos.discovery.instance-enabled=true","-Dspring.cloud.nacos.discovery.cluster-name=gateway-RVC962","-Dspring.application.name=gateway","-Dspring.cloud.inetutils.preferredNetworks=10.0.2","-jar","app.jar",">gateway.log" ]

ENTRYPOINT [ "java" ]

Is that so?
